xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +135,793 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
+
+	closed *grpcsync.Event
+	cc     balancer.ClientConn
+	bOpts  balancer.BuildOptions
+	logger *grpclog.PrefixLogger
+
+	// childMu protects child and also updates to the child (to uphold the
+	// balancer.Balancer API guarantee of synchronous calls). It also protects
+	// against run() reading that the child is not nil for SubConn updates, and
+	// then UpdateClientConnState or Close writing to the the child.
+	childMu sync.Mutex
+	child   balancer.Balancer
+
+	// mu guards access to a lot of the core LB Policy State. It also prevents
+	// intersplicing certain operations.
+	//
+	// ex 1: interval timer goes off, outlier detection algorithm starts running
+	// based on knobs in cfg. in the middle of running the algorithm, a
+	// ClientConn update comes in and writes to cfg. This causes undefined
+	// behavior for the interval timer algorithm.
+	//
+	// ex 2: Updating the addrs map from UpdateAddresses in the middle of
+	// running the interval timer algorithm which uses addrs heavily. This will
+	// cause undefined behavior for the interval timer algorithm.
+	mu             sync.Mutex
+	addrs          map[string]*addressInfo
+	cfg            *LBConfig
+	scWrappers     map[balancer.SubConn]*subConnWrapper
+	timerStartTime time.Time
+	intervalTimer  *time.Timer
+
+	scUpdateCh     *buffer.Unbounded
+	pickerUpdateCh *buffer.Unbounded
+}
+
+// noopConfig returns whether this balancer is configured with a logical no-op
+// configuration or not.
+func (b *outlierDetectionBalancer) noopConfig() bool {
+	return b.cfg.SuccessRateEjection == nil && b.cfg.FailurePercentageEjection == nil
+}
+
+func (b *outlierDetectionBalancer) UpdateClientConnState(s balancer.ClientConnState) error {
+	lbCfg, ok := s.BalancerConfig.(*LBConfig)
+	if !ok {
+		b.logger.Errorf("received config with unexpected type %T: %v", s.BalancerConfig, s.BalancerConfig)
+		return balancer.ErrBadResolverState
+	}
+
+	// Reject whole config if any errors, don't persist it for later
+	bb := balancer.Get(lbCfg.ChildPolicy.Name)
+	if bb == nil {
+		return fmt.Errorf("outlier detection: child balancer %q not registered", lbCfg.ChildPolicy.Name)
+	}
+
+	if b.child == nil || b.cfg.ChildPolicy.Name != lbCfg.ChildPolicy.Name {
+		b.childMu.Lock()
+		if b.child != nil {
+			b.child.Close()
+		}
+		b.child = bb.Build(b, b.bOpts)
+		b.childMu.Unlock()
+	}
+
+	b.mu.Lock()
+	b.cfg = lbCfg
+
+	// When the outlier_detection LB policy receives an address update, it will
+	// create a map entry for each subchannel address in the list, and remove
+	// each map entry for a subchannel address not in the list.
+	addrs := make(map[string]bool, len(s.ResolverState.Addresses))
+	for _, addr := range s.ResolverState.Addresses {
+		addrs[addr.Addr] = true
+		if _, ok := b.addrs[addr.Addr]; !ok {
+			b.addrs[addr.Addr] = newAddressInfo()
+		}
+	}
+	for addr := range b.addrs {
+		if !addrs[addr] {
+			delete(b.addrs, addr)
+		}
+	}
+
+	// When a new config is provided, if the timer start timestamp is unset, set
```

Comment:
A lot of comments in here are just stating exactly what is happening. Could you please revisit them and prune them a bit, and only retain things which can't be gathered easily by simply reading the code.

Reply:
Yeah, good point. I like quoting the designs, but you're right, the spec is telling you what the code does and now the code should explain what is happening by itself. Deleted most references to A50 throughout this file.